### ISSUE TITLE
Some maker specific tags cannot be deleted

### ISSUE THREAD

**Issue by norbertwg** [Exiv2/exiv2#2379]
#### **Describe the bug**
Tried to delete the following tags, but they were not deleted:
Exif.CanonPi.AFPointsUsed in [exiv2-canon-eos-300d.jpg](https://github.com/Exiv2/exiv2/blob/main/test/data/exiv2-canon-eos-300d.jpg)
Exif.CanonSi.SubjectDistance in [20220610_MG_7238.exv](https://github.com/Exiv2/exiv2/blob/main/test/data/20220610_MG_7238.exv)
Exif.CanonSi.AFPointUsed in [20220610_MG_7238.exv](https://github.com/Exiv2/exiv2/blob/main/test/data/20220610_MG_7238.exv)
Exif.CanonCs.LensType in [20220610_MG_7238.exv](https://github.com/Exiv2/exiv2/blob/main/test/data/20220610_MG_7238.exv)
Exif.NikonIi.ISO in [CH0_0174.exv](https://github.com/Exiv2/exiv2/blob/main/test/data/CH0_0174.exv)

#### **To Reproduce**
Steps to reproduce the behavior (for the first tag listed above, for the others analog):
$ exiv2 -K Exif.CanonPi.AFPointsUsed  exiv2-canon-eos-300d.jpg
Exif.CanonPi.AFPointsUsed                    Short       1  center

$ exiv2 -M"del Exif.CanonPi.AFPointsUsed"  exiv2-canon-eos-300d.jpg

$ exiv2 -K Exif.CanonPi.AFPointsUsed  exiv2-canon-eos-300d.jpg
Exif.CanonPi.AFPointsUsed                    Short       1

Branch: main [b1e2cc6](https://github.com/Exiv2/exiv2/commit/b1e2cc6129516f44b30956f3dc6ec2765b7ae475)

#### **Expected behavior**
After deletion, tags should not be listed any longer, even not with a blank value.

#### **Desktop (please complete the following information):**
-  OS and version: Windows 11
-  Exiv2 version and source: exiv2 1.0.0.9
-  Compiler and version: MSVC 2022
-  Compilation mode and/or compiler flags: Release

**Comment by postscript-dev** [comment]
@norbertwg 
> Tried to delete the following tags, but they were not deleted:

As it happens, I have recently been looking at the same thing. I noticed the following when working on the Sony tags:
1. Tags in the main section of the makernotes (e.g., `Exif.Sony1.Quality`) can be deleted.
2. New tags added to the main section of the makernotes  (e.g., `Exif.Sony1.WhiteBalance`) are added at the end of the makernotes. If displaying all the Exif tags in order, the `Exif.Sony1` tags are now in two parts with other Sony groups in between.
3. Tags in sub-groups of the makernotes (e.g., `Exif.Sony2010e.SonyDateTime`) cannot be removed but if they are deleted, the buffer is zeroed.
4. Sub-group tags in the makernotes (e.g., `Exif.Sony2010e.SonyDateTime`) are fixed in size. When modifying, string/undefined types are padded with zeros.
5. Sub-groups in the makernotes (e.g., `Exif.Sony2010e`) cannot be deleted.

I think that I remember a related discussion somewhere about this. Most tags in the makernotes are unknown and some of them may act as offsets (e.g. the location of a preview image). If we delete/modify tags, then the data layout of the makernotes is changed and any unknown offsets could become invalid.

I am currently filling gaps in [exiv2.md](https://github.com/Exiv2/exiv2/blob/main/exiv2.md) and have added my above observations.

**Comment by norbertwg** [comment]
@postscript-dev 
Seems, that fixing this issue (and my other two issues about deleting tags: [#2357 After deleting Exif.Nikon3.ISOSpeed, Exif.NikonSi02xx.ISO instead of Exif.NikonSiD300a.ISO is shown](https://github.com/Exiv2/exiv2/issues/2357) and [#2356 After deleting Exif.Photo.ExposureProgram, maker tags from Olympus disappeared as well](https://github.com/Exiv2/exiv2/issues/2356)) will not be easy. So I would like to add some thoughts about deleting tags.

I found the issues when I tried to prepare test files to test easyaccess. There I needed to have files having one tag, but not having some others. This reason to delete tags is not a frequent one. The usual reasons to delete tags are probably either to remove confidential data like serial number or to remove all tags to decrease file size. Deleting all tags is not the topic here. When the reason is to get rid of confidential data, instead of really deleting the tag, blanking resp. zeroing the value could be acceptable (and as far as I know this would apply only for maker specific tags). There remains a small risk, that a zero value for some tags may have a specific meaning or in other words, the value zero (and if available its translation) has an other meaning for the user than not seeing the tag at all.

To summarize it: Considering the potential effort of fixing the issue and how often it is probably needed, blanking or zeroing maker specific tags is an acceptable alternative to really delete it and better than having side effects as indicated in issue [#2356 After deleting Exif.Photo.ExposureProgram, maker tags from Olympus disappeared as well](https://github.com/Exiv2/exiv2/issues/2356) .

**Comment by postscript-dev** [comment]
@norbertwg 

> There remains a small risk, that a zero value for some tags may have a specific meaning or in other words, the value zero (and if available its translation) has an other meaning for the user than not seeing the tag at all.

I noticed this problem too. In some situations, zeros mean `n/a` but this is context dependant.

> To summarize it: Considering the potential effort of fixing the issue and how often it is probably needed, blanking or zeroing maker specific tags is an acceptable alternative to really delete it

I don't know much about deleting makernote tags as it is part of the code I haven't investigated yet. I will try and take a look into the subject but it may take a few days.

If you want to, you can read more about preview images and many other subjects in [Robin's exiv2 book](https://exiv2.org/book/index.html).

**Comment by postscript-dev** [comment]
@norbertwg 
> I don't know much about deleting makernote tags as it is part of the code I haven't investigated yet. I will try and take a look into the subject but it may take a few days.

Sorry for the delay in replying. I have been looking at the issue and have a couple of observations.

As ExifTool has more mature makernote support, I decided to investigate how they cope with the deletion issue. Their manpage explains:
```
         2) In general, MakerNotes tags are considered "Permanent", and may
         be edited but not created or deleted individually. This avoids many
         potential problems, including the inevitable compatibility problems
         with OEM software which may be very inflexible about the
         information it expects to find in the maker notes.
```

When I tried the [Exiv2 deletion rules](https://github.com/Exiv2/exiv2/issues/2379#issuecomment-1272633967) using ExifTool, the results were similar except:
1. Makernote tags *cannot be deleted (i.e., removed)*.
2. Deleting makernote string tags (ExifTool: `string` type; Exiv2: `Ascii` type) in the main section (e.g., `Exif.Sony1.CreativeStyle`) *truncates the value to one null character*.
3. Deleting any other makernote tags (e.g., `Exif.Sony1.Quality`) preserves the size but zeros the data.
4. Makernote tags *cannot be added*.

After looking at this, I am wondering if Exiv2's makernote tag policy is correct.

As for the issue of the makernote preview image offsets, I see that Exiv2 does not extract larger preview images (except for BMFF-based files which use a different system). The smaller preview images are stored in specific makernote tags (e.g., `Exif.PanasonicRaw.PreviewImage`) but the larger images can be stored elsewhere. When I experimented by deleting tags in a Sony file with Exiv2, ExifTool was still able to correctly extract all the preview images. I didn't try preview images for other camera manufacturers.

I don't think that I understand this topic enough to decide what to do. If I find time, I will investigate more.
